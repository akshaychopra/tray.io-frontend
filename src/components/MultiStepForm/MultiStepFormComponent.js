import FormStep from "../MultiStepForm/FormStep";
import { useState } from "react";
import Stepper from "@material-ui/core/Stepper";
import Step from "@material-ui/core/Step";
import StepLabel from "@material-ui/core/StepLabel";

import { makeStyles } from "@material-ui/core/styles";
import { Card, CardContent, createMuiTheme, ThemeProvider } from "@material-ui/core";

const defaultTheme = createMuiTheme();
// custom theme for multistep forms
const theme = createMuiTheme({
  spacing: 2,
  color: {
    green: defaultTheme.palette.success.main,
  },
  border: {
    normal: `1px solid #CCC`,
    error: `1px solid ${defaultTheme.palette.error.light}`,
  },
});

const useStyles = makeStyles((theme) => ({
  root: {
    margin: theme.spacing(4),
  },
}));

/**
 * Component to create a multi step form
 * @param {JSON} config - forms configuration parameters
 */
const MultiStepFormComponent = ({ config }) => {
  const classes = useStyles();
  // defaults to 0 step and first page
  const [step, setStep] = useState(0);
  const [stepConfig, setStepConfig] = useState(config.pages[0]);

  const onStepSubmit = (step, formData) => {
    // if not at last page move to next page
    if (!(config.pages.length - 1 === step)) {
      setStep(step + 1);
      setStepConfig(config.pages[step + 1]);
    }
    // if the config tells you to explicitly submit or it is the last page then submit form
    if (config.pages[step].submitData || config.pages.length - 1 === step) {
      config.onSubmit(formData);
    }
  };
  return (
    <ThemeProvider theme={theme}>
      <Card className={classes.root}>
        <CardContent>
          <Stepper activeStep={step}>
            {config.pages.map((page) => (
              <Step key={page.name}>
                <StepLabel>{page.name}</StepLabel>
              </Step>
            ))}
          </Stepper>
          <FormStep config={stepConfig} step={step} submit={onStepSubmit} />
        </CardContent>
      </Card>
    </ThemeProvider>
  );
};

export default MultiStepFormComponent;
