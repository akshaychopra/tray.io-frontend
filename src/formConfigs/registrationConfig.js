import SuccessComponent from "../components/SuccessComponent";

const VALID_EMAIL_REGEX = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/gi;
const VALID_PASSWORD_REGEX = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)[a-zA-Z\d\D]{8,}$/g;

export const REGISTRATION_FORM = {
  pages: [
    {
      name: "User",
      fields: [
        {
          label: "name",
          name: "name",
          placeholder: "Name",
          type: "text",
          validation: {
            required: { message: "Please enter a name" },
          },
        },
        {
          label: "role",
          name: "role",
          placeholder: "Role",
          type: "text",
        },
        {
          label: "email",
          name: "email",
          placeholder: "Email",
          type: "text",
          validation: {
            regex: {
              regex: VALID_EMAIL_REGEX,
              message: "Please enter valid a email",
            },
          },
        },
        {
          label: "password",
          name: "password",
          placeholder: "Password",
          type: "password",
          validation: {
            regex: {
              regex: VALID_PASSWORD_REGEX,
              message:
                "The password needs to be a mix of at least one uppercase letter, one lowercase letter, one number and be greater than 9 characters",
            },
          },
        },
      ],
      submit: "Submit",
    },
    {
      name: "Privacy",
      fields: [
        {
          label: "updates",
          name: "updates",
          text: "Receive updates about the Tray.io product by email",
          type: "checkbox",
        },
        {
          label: "communication",
          name: "communication",
          text: "Receive communication about other products by Tray.io product by email",
          type: "checkbox",
        },
      ],
      submit: "Submit",
      submitData: true, // flag submit data at any given point
    },
    {
      name: "Done",
      component: <SuccessComponent />,
    },
  ],
  onSubmit: (values) => {
    // TODO do something with values
    console.log(values);
  },
};
