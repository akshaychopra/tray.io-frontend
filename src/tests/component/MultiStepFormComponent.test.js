import React from "react";
import { render, fireEvent } from "@testing-library/react";
import MultiStepFormComponent from "../../components/MultiStepForm/MultiStepFormComponent";
import "@testing-library/jest-dom/extend-expect";
import { REGISTRATION_FORM } from "../../formConfigs/registrationConfig";
import renderer from "react-test-renderer";

test("Renders registration form", () => {
  const tree = renderer.create(<MultiStepFormComponent config={REGISTRATION_FORM} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("Fills out the registration form and make sure the success result is printed", async () => {
  const testObj = {
    name: "testname",
    email: "testemail@test.com",
    password: "aaaAAA111",
    role: "SD",
  };
  const { getByPlaceholderText, getByText } = render(<MultiStepFormComponent config={REGISTRATION_FORM} />);
  const nameInput = getByPlaceholderText("Name");
  fireEvent.change(nameInput, { target: { value: testObj.name } });
  const emailInput = getByPlaceholderText("Email");
  fireEvent.change(emailInput, { target: { value: testObj.email } });
  const roleInput = getByPlaceholderText("Role");
  fireEvent.change(roleInput, { target: { value: testObj.role } });
  const passwordInput = getByPlaceholderText("Password");
  fireEvent.change(passwordInput, { target: { value: testObj.password } });
  fireEvent.click(getByText("Submit"));
  fireEvent.click(getByText("Submit"));
  const linkElement = getByText(/Please verify your/i);
  expect(linkElement).toBeInTheDocument();
});

// mock console log
