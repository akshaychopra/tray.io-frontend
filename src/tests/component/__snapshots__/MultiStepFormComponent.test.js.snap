// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders registration form 1`] = `
<div
  className="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardContent-root"
  >
    <div
      className="MuiPaper-root MuiStepper-root MuiStepper-horizontal MuiPaper-elevation0"
    >
      <div
        className="MuiStep-root MuiStep-horizontal"
      >
        <span
          className="MuiStepLabel-root MuiStepLabel-horizontal"
        >
          <span
            className="MuiStepLabel-iconContainer"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiStepIcon-root MuiStepIcon-active"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="12"
                cy="12"
                r="12"
              />
              <text
                className="MuiStepIcon-text"
                textAnchor="middle"
                x="12"
                y="16"
              >
                1
              </text>
            </svg>
          </span>
          <span
            className="MuiStepLabel-labelContainer"
          >
            <span
              className="MuiTypography-root MuiStepLabel-label MuiStepLabel-active MuiTypography-body2 MuiTypography-displayBlock"
            >
              User
            </span>
          </span>
        </span>
      </div>
      <div
        className="MuiStepConnector-root MuiStepConnector-horizontal Mui-disabled"
      >
        <span
          className="MuiStepConnector-line MuiStepConnector-lineHorizontal"
        />
      </div>
      <div
        className="MuiStep-root MuiStep-horizontal"
      >
        <span
          className="MuiStepLabel-root MuiStepLabel-horizontal Mui-disabled"
        >
          <span
            className="MuiStepLabel-iconContainer"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiStepIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="12"
                cy="12"
                r="12"
              />
              <text
                className="MuiStepIcon-text"
                textAnchor="middle"
                x="12"
                y="16"
              >
                2
              </text>
            </svg>
          </span>
          <span
            className="MuiStepLabel-labelContainer"
          >
            <span
              className="MuiTypography-root MuiStepLabel-label MuiTypography-body2 MuiTypography-displayBlock"
            >
              Privacy
            </span>
          </span>
        </span>
      </div>
      <div
        className="MuiStepConnector-root MuiStepConnector-horizontal Mui-disabled"
      >
        <span
          className="MuiStepConnector-line MuiStepConnector-lineHorizontal"
        />
      </div>
      <div
        className="MuiStep-root MuiStep-horizontal"
      >
        <span
          className="MuiStepLabel-root MuiStepLabel-horizontal Mui-disabled"
        >
          <span
            className="MuiStepLabel-iconContainer"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiStepIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <circle
                cx="12"
                cy="12"
                r="12"
              />
              <text
                className="MuiStepIcon-text"
                textAnchor="middle"
                x="12"
                y="16"
              >
                3
              </text>
            </svg>
          </span>
          <span
            className="MuiStepLabel-labelContainer"
          >
            <span
              className="MuiTypography-root MuiStepLabel-label MuiTypography-body2 MuiTypography-displayBlock"
            >
              Done
            </span>
          </span>
        </span>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    >
      
      <div
        className="makeStyles-fieldContainer-2"
      >
        <input
          className="makeStyles-field-3 "
          label="name"
          name="name"
          onChange={[Function]}
          placeholder="Name"
          type="text"
        />
        <div
          className="makeStyles-fieldText-5"
        />
        <div
          className="makeStyles-fieldError-6"
        />
      </div>
      <div
        className="makeStyles-fieldContainer-2"
      >
        <input
          className="makeStyles-field-3 "
          label="role"
          name="role"
          onChange={[Function]}
          placeholder="Role"
          type="text"
        />
        <div
          className="makeStyles-fieldText-5"
        />
        <div
          className="makeStyles-fieldError-6"
        />
      </div>
      <div
        className="makeStyles-fieldContainer-2"
      >
        <input
          className="makeStyles-field-3 "
          label="email"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="text"
        />
        <div
          className="makeStyles-fieldText-5"
        />
        <div
          className="makeStyles-fieldError-6"
        />
      </div>
      <div
        className="makeStyles-fieldContainer-2"
      >
        <input
          className="makeStyles-field-3 "
          label="password"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
        />
        <div
          className="makeStyles-fieldText-5"
        />
        <div
          className="makeStyles-fieldError-6"
        />
      </div>
      <div
        className="makeStyles-action-7"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Submit
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
